Modeling designs page review (#451)
neo4j/docs-getting-started: modules/ROOT/pages/data-modeling/modeling-designs.adoc

Now there are no copies or duplications.

=== Organizing data

Intermediate nodes can also help organize structures.
In this example, Sarah sent the same email message to several people:

image::sarah-emailed.svg[Graph showing how a user named Sarah emailed several other users,width=300,role=popup]

If every `EMAILED` relationship included a property with the content of the message, two problems can arise:

* *Sarah’s node is becoming very dense*: For every recipient of a message she adds, her node gains another relationship.
* *It's expensive to retrieve the content of the email*: With the data modeled like this, it's very expensive to determine who in Sarah’s recipient network has received a given message by searching for the content in multiple 'EMAILED' relationships.

By adding intermediate nodes to represent each email message, Sarah's node has only one relationship per email message, regardless of the number of recipients:

image::sarah-emailed-intermediate.svg[Updated graph with intemediate nodes for emails sent by Sarah to several other users,width=300,role=popup]

> With this model, you can find the recepients by locating the specific `Email` node that now contains the content of the message in the `content` property, and then see which users are connected to it via `TO` relationships.

Author: It could be through an email ID instead of the content. Do you think it's necessary to add this?

While both models use a gather-and-inspect approach, the scope of the problem is reduced significantly after the refactoring.
In the first iteration, if you want to see who received a certain email, you need to find all users connected to Sarah via the `EMAILED` relationship.
In the second iteration, you only need to locate the correct `Email` node, then traverse from it to all of the connected recipients.

In summary, you're likely to find many uses for intermediate nodes during refactoring since you rarely recognize the need for them at the outset of the data modeling.

== Linked list

Linked lists are commonly used in computer science and they are particularly useful whenever the sequence of objects matters.
In this data structure, a simple-linked list is where each node links to the next node only:

image::simple-linked-list.svg[Episodes of Dr Who linked in sequence with next relationships,width=500,role=popup]

> In a double-linked list, each node links both to the next and the previous node:

Reviewer: ```suggestion
In a *double-linked list*, each node links both to the next and the previous node:
```

[[modeling-designs]]
= Modeling designs
:tags: graph-modeling, data-model, schema, model-design, modeling-decisions
:description: This page features examples of graph data modeling patterns and designs that are commonly used with Neo4j.

This page features examples of graph data modeling patterns and designs that are commonly used with Neo4j.
The purpose is to get an overview of the options available for building graph data models and how known strategies can be adapted to your project.

== Intermediate nodes

Intermediary nodes are nodes that contain data that need to be in the graph but don't seem to fit neatly into the initial model.

Sometimes you need to convey a lot of information in a relationship.
In a mathematical graph, this can be solved with a *hyperedge*, i.e. a relationship that connects more than two nodes.
This is not supported in Neo4j but can be solved by using an intermediary node.

For example, consider a person who works at a company and you need to convey information about their role:

> image::hyperedge.svg[An example of a hyperedge in which a relationship is connected to two nodes, a feature not available in Neo4j,width=400,role=popup]

Author: This should have been fixed yet. Is this the image that you are seeing? https://github.com/neo4j/docs-getting-started/pull/451/files#diff-c7e86415c640b7192119c42108fda32b36bb6cba201e6f5ee9d80b1a63a0b021

With this model, you can find the recepients by locating the specific `Email` node that now contains the content of the message in the `content` property, and then see which users are connected to it via `TO` relationships.

While both models use a gather-and-inspect approach, the scope of the problem is reduced significantly after the refactoring.
In the first iteration, if you want to see who received a certain email, you need to find all users connected to Sarah via the `EMAILED` relationship.
In the second iteration, you only need to locate the correct `Email` node, then traverse from it to all of the connected recipients.

In summary, you're likely to find many uses for intermediate nodes during refactoring since you rarely recognize the need for them at the outset of the data modeling.

== Linked list

Linked lists are commonly used in computer science and they are particularly useful whenever the sequence of objects matters.
In this data structure, a simple-linked list is where each node links to the next node only, whereas in a double-linked list, each node links both to the next and the previous node.

> Neo4j *does not* support double-linked lists.

Reviewer 2: Technically, nothing prevents from doing this. 
However this is redundant info, as one relationship implies the other. They are semantically equivalent.
There is no performance difference traversing in both directions.
I would rephrase at this is not recommended.

a| image::double-linked-list.svg[Episodes of Dr Who doubly linked with next and previous relationships, an incorrect modeling design in Neo4j,role=popup]
|===

=== Interleaved linked list

A list of items is not always sequential.
When nodes are connected in a nonlinear fashion, they are referred to as an interleaved linked list.

Interleaved lists are used when you want to sequence a set of list based on context, not necessarily a sequence in time.
This example combines a linked list with an interleaved linked list of Dr. Who episodes:

image::interleaved-list.svg[Example of interleaved list connecting Dr Who episodes with next and next in production relationships,width=700,role=popup]

The order in which TV episodes are aired is often different than the order in which they are produced.
This example contains five episodes of Dr. Who from season 12 and it shows:

* The order in which the episodes were aired using the `NEXT` relationship and through a simple-linked list.
* The order in which the episodes were produced using the `NEXT_IN_PRODUCTION` relationship, which creates an interleaved linked list.

> It is not a linear list, as it goes 1, 3, 2, 5, 4.

Author: What would be a non-linear example then? I'm struggling to understand 😅

The parent almost always points to the first item in the sequence, using an appropriately named relationship. 
Sometimes, another relationship points to the last item in a list.

In this example, you can see a `FIRST` and a `LAST` relationship, referring to their places in the sequence:

image::head-tail-list.svg[List of five Dr Who episodes from season twelve, showing the first and the last through a relationship with the season node,width=700,role=popup]

Some implementations also have a "progress" pointer that is used to keep track of the current node of interest.
This can be done through a relationship, as such:

image::latest-aired.svg[The latest aired episode is pointed with the help of a latest aired relationship coming from the season node,width=700,role=popup]

The progress pointer here is the `LATEST_AIRED` relationship and it shows which was the most recently aired episode (i.e. "The Ark in Space").
When the `NEXT` episode ("The Sontaran Experiment") airs, the relationship is updated by deleting the current one and creating a new `LATEST_AIRED` pointer, so that it always points to the current item.

> == Timeline tree

Reviewer 2: The whole time tree section needs to be removed.
It used to be useful when neo4j did not have indexes on dates. This is not the case anymore. This pattern is obsolete.
Not sure why it ended up in the PS KB in the first place.

Timeline trees are useful when your queries are time-related.
To create a timeline tree, you can use an anchor, which is a navigational aid that you add when querying or when there are periods of time that are of interest relative to a query.
Anchors can be a starting/parent node or an entity that you can use as a reference point (e.g. a date).

The division and length of the time periods are set based on the context of the question you need to answer.
On the top of the timeline tree is the "all time node" that represents the entire timeline.

> The timeline is then divided into relevant time periods, represented by the nodes below the all time node.

Reviewer: I think we're missing an important piece of information here. This (from the KB article):
_The rest of the data nodes, which are non-time data nodes, are the nodes that contain the important pieces data in the graph. These nodes link into the timeline tree at the appropriate leaf node._

This example features subdivisions for each branch in years, months, and days, and how they connect to `Purchase` nodes:

image::timeline-tree.svg[Graph with two different timeslines divided in years, months, and days and what purchases are connected to these dates,width=600,role=popup]

> This timeline tree structure shows how otherwise expensive queries can be answered more quickly.

Reviewer: ```suggestion
This timeline tree structure shows how you can quickly answer otherwise expensive questions.
```

For instance, if you want to find all purchases that happened at a given time period, such as every purchase in the month of December 2012, you only need to:

. Navigate from the 2012 `Year` node to December (`Month` node with the property value "12").
. Fetch everything from the connected leaf nodes (that is, nodes with no descendants) under that branch.

If you were, instead, working with a data model that uses time as the navigator instead of as the anchor, this same question would require a great deal of property lookups and a lot of inefficient gather-and-inspect.

> == Time-bound data

Reviewer 2: same as for time tree, this is not true anymore now that date indexes are available

> If you were, instead, working with a data model with no timeline tree, this same question could require a great deal of property lookups and a lot of inefficient gather-and-inspect.

Reviewer: ```suggestion
If you were, instead, working with a data model that uses time as the navigator instead of as the anchor, this same question would require a great deal of property lookups and a lot of inefficient gather-and-inspect.
```

image::airport.svg[Example graph connecting airport nodes through a flying to relationship,width=500,role=popup]

This could be a first iteration showing what flights happen between airports.
Now, suppose you want to know about a particular airline and how many flights it had on a specific date.
The described model is not very efficient as you would need to scan all properties of the `FLYING_TO` relationship and, in reality, airports have multiple flights operated by many different airlines everyday.

So instead of performing an expensive query, you could refactor the graph and add xref:#_intermediate_nodes[intermediate nodes].
In this case, you can create an intermediate node `Flight` between the `Airport` nodes, and another one, `AirportDay`, between the `Airport` and `Flight` nodes:

image::flight.svg[Refactored graph with intermediate nodes,width=600,role=popup]

With this xref:data-modeling/tutorial-refactoring.adoc[refactored design], you can look only at the relationships for dates you care about and then scan for the relevant airline through the `airline` property in the `Flight` node.

This is an improvement, but the model is still not optimal and should be iterated on.

> When changes are done to your model, it is important to keep track of them by versioning your model.

Reviewer: Why is it important to keep track of versions of your model?

> // Content on versioning is WIP.

Author: Please refer to https://github.com/neo4j/docs-getting-started/pull/465

Take this example:

image::multiple-structures.svg[Example graph using multiple structures to create a data model with timeline tree and employment events,width=600,role=popup]

This graph features:

* A xref:#_timeline_tree[timeline tree] which is divided into years and months subsequently at the top.
* Employment events are represented as `Employment` nodes, which are connected to the timeline and show the starting and the ending months.
* The `Employment` nodes are xref:#_intermediate_nodes[intermediate nodes] that connect people, companies, and job roles. 
In this case, the intermediate nodes are serving the dual purpose of reducing the density of `Person` nodes, and providing an attachment point for the timeline tree.
* Employment events that are organized into xref:#_linked_list[linked lists] for each person and represent a sequence of events in an individual’s career.
The graph shows overlaps in people's employment history.

The example shows how different structures can be combined into one graph, and how different types of questions can be answered by one single graph if the modeling is done efficiently.
It also shows the many ways nodes and structures can span out.

> === Monopartite

Reviewer 2: I'm not sure monopartite / bipartite / multipartite brings value here.
I don't see this as a pattern as people will choose one or the other representation according to the use case. There is no benefit of one over the others. It would see this more as a GDS topic, where a given shape must be used according to the algorithm.

> === Monopartite

Reviewer: This feels a little abrupt? Maybe it can be introduced with a little sentence about different graph structures or something?

image::multipartite.svg[Example of a multipartite structure featuring multiple independent sets of nodes, labels, and different relationships,width=600,role=popup]

== Intermediate nodes

Intermediate nodes are nodes that are needed, but often aren’t identified in the initial data model.
They are nodes that contain data that needs to be in the graph, but don't seem to fit in the beginning of the modeling process.

This kind of situation often occurs when you use a single relationship to connect *more* than two nodes.
While this is an acceptable strategy in mathematical graphs, which can use hyperedges, it is not supported in Neo4j.

One way to identify that you’re dealing with a potential hyperedge is when you want to locate something in a relationship rather than a node. 
For example, here there is a person who works at a company and there is information about their role:

image::hyperedge.svg[An example of a hyperedge in which a relationship is connected to two nodes, a feature not available in Neo4j,width=400,role=popup]

In a mathematical graph, you could use the same relationship `WORKED_AT` to connect the `Person` node with both `Role` and `Company` nodes.

> However, in Neo4j, you can't do it.

Author: The example shows that you can't create a relationship coming from a relationship in Neo4j, thus the need for an intermediate node. I think the example in GraphAcademy is also correct, but different. I don't know which one is the most relevant for users. Maybe when @nmervaillie reviews this PR he could give us some insight?

> Instead, you could either turn the `Role` node into a property of the `WORKED_AT` relationship or use an *intermediate node* between the `Person`, `Company`, and `Role` nodes:

Author: I'm not sure if I understand. How would you suggest the example be?

In summary, you're likely to find many uses for intermediate nodes, but you might rarely recognize the need for them at the outset of the data modeling.
When refactoring and re-imagining the graph, however, they prove to be very useful.

== Linked list

Linked lists are commonly used in computer science and they are particularly useful whenever the sequence of objects matters.
Despite being a good practice to use verbs when creating a relationship, it is very common in this case to connect sequential items using terms such as "next" and "previous" instead.

With this strategy, you can have simple- or double-linked lists, but *you should never use a double-linked list in Neo4j*.
For example:

[cols="<,<",options="header"]
|===

a| **Correct: simple-linked list**
a| **Incorrect: double-linked list**

a| image::simple-linked-list.svg[Episodes of Dr Who linked in sequence with next relationships,role=popup]
a| image::double-linked-list.svg[Episodes of Dr Who doubly linked with next and previous relationships, an incorrect modeling design in Neo4j,role=popup]
|===

=== Interleaved linked list

There are different ways to sequence a list of items.

> When nodes are connected in a nonlinear fashion, they are referred to as an interleaved linked list.

Reviewer 2: Not sure I get the question. 
I don't see this as linear or not. Both relationship types just represent different information, sharing pieces of content

This timeline tree structure shows how otherwise expensive queries can be answered more quickly.
For instance, if you want to find all purchases that happened at a given time period, such as every purchase in the month of December 2012, you only need to:

. Navigate from the 2012 `Year` node to December (`Month` node with the property value "12").
. Fetch everything from the connected leaf nodes under that branch.

If you were, instead, dealing with time as a navigator, but not as the anchor, and you wanted to answer the same question, you would need to:

. Start by finding the event of a specific purchase.
. Find all other purchases within a year.
. Start a leaf node for the original purchase.
. Traverse into the timeline tree, going up to the year of interest.
. Go back down the tree, fetching all relevant data while descending it. 

Without a timeline tree in the graph, a same question could require a great deal of property lookups and a lot of inefficient gather-and-inspect.

== Time-bound data

One way to model time-specific data and relationships is by including data in the relationship type.

> Because Neo4j is optimized specifically for traversing relationships between entities, you can often improve query performance by specifying a date as the relationship type and only traversing particular dated relationships.

Reviewer: ```suggestion
Because Neo4j is optimized for traversing relationships between entities, you can often improve query performance by using a date as the relationship type and only traversing relationships with a particular date.
```

A common example is modeling airline flights:

image::airport.svg[Example graph connecting airport nodes through a flying to relationship,width=500,role=popup]

> This could be a first iteration showing what flights happen between airports.

Reviewer: ```suggestion
This could be a first iteration showing what flights there are between airports.
```

> Now, suppose you want to know about an airline in particular and how many flights it performed during a specific date.

Reviewer: ```suggestion
Now, suppose you want to know about a particular airline and how many flights it flew on a specific date.
```

> This particular model is not very efficient as you would need to scan for this information through all properties of the `FLYING_TO` relationship and, in reality, airports perform multiple flights from several airlines everyday.

Reviewer: ```suggestion
The described model is not very efficient as you would need to scan all properties of the `FLYING_TO` relationship and, in reality airports have multiple flights from many different airlines everyday.
```

> So instead of performing an expensive query, you could refactor the graph to add xref:#_intermediate_nodes[intermediate nodes].

Reviewer: ```suggestion
So instead of performing an expensive query, you could refactor the graph and add xref:#_intermediate_nodes[intermediate nodes].
```

In this case, you can create an intermediate node `Flight` between the `Airport` nodes, and another one, `AirportDay`, between the `Airport` and `Flight` nodes:

image::flight.svg[Refactored graph with intermediate nodes,width=600,role=popup]

> With this xref:data-modeling/graph-model-refactoring.adoc[refactored design], you can look only at the relationships for dates you care about and then scan for the relevant airline through the `Airline` property in the `Flight` node.

Reviewer: ```suggestion
With this xref:data-modeling/graph-model-refactoring.adoc[refactored design], you can look only at the relationships for dates you care about and then scan for the relevant airline through the `airline` property in the `Flight` node.
```
Not sure if the link to refactoring should be this far down the page, it could probably be added earlier on.

> Despite the improvement, this model is still not the most optimal design, but it is nevertheless an iteration.
> When such changes are performed, it can be important to keep track of them by versioning your model.

Reviewer: ```suggestion
This is an improvement, but the model is still not optimal and should be iterated on.
When changes are done to your model, it is important to keep track of them by versioning your model.
```
